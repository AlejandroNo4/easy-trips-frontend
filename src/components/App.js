import { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useNavigate, Outlet } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faBars, faHome } from '@fortawesome/free-solid-svg-icons';
import fetchingGet from '../api/fetchingGet';
import UserPanel from './UserPanel';

function App() {
  const userState = useSelector((state) => state.UIReducer);
  const [showPanel, updatePanel] = useState('d-none');
  const navigate = useNavigate();
  const dispatch = useDispatch();

  useEffect(() => {
    const url = 'logged_in';
    const type = 'UI';
    fetchingGet({
      dispatch,
      url,
      type,
    });
  }, []);

  const loginHandler = () => {
    navigate('/login');
  };

  const signUpHandler = () => {
    navigate('/sign-up');
  };

  const homeHandler = () => {
    navigate('/');
  };

  const toggleUserPanel = () => {
    if (showPanel === 'd-none') {
      updatePanel('d-flex');
    } else {
      updatePanel('d-none');
    }
  };

  if (userState.loading === true) {
    return <h1>------LOADING USER...------</h1>;
  }
  if (userState.user.logged_in === true && userState.loading === false) {
    return (
      <div className="d-flex">
        <UserPanel display={showPanel} />
        <div>
          <div className="nav-bar-mobile d-flex space-between">
            <FontAwesomeIcon icon={faBars} onClick={toggleUserPanel} />
            <h1 className="main-title">Easy Trips</h1>
            <FontAwesomeIcon icon={faHome} onClick={homeHandler} />
          </div>
        </div>
        <Outlet />
      </div>
    );
  }
  return (
    <div className="app">
      <h1>Connected!</h1>
      <h2>Status: NOT logged in</h2>
      <button onClick={loginHandler} type="button">
        ---Login---
      </button>
      <button onClick={signUpHandler} type="button">
        ---Sign Up---
      </button>
    </div>
  );
}

export default App;
