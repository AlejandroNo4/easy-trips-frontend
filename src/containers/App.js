import React, { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useNavigate, Outlet } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faBars, faHome } from '@fortawesome/free-solid-svg-icons';
import fetchingGet from '../api/fetchingGet';
import UserPanel from '../components/UserPanel';
import NoSession from '../components/NoSession';

function App() {
  const userState = useSelector((state) => state.UIReducer);
  const [showPanel, updatePanel] = useState('d-none');
  const navigate = useNavigate();
  const dispatch = useDispatch();

  useEffect(() => {
    const url = 'logged_in';
    const type = 'UI';
    fetchingGet({
      dispatch,
      url,
      type,
    });
  }, []);

  const loginHandler = () => {
    navigate('/login');
  };

  const signUpHandler = () => {
    navigate('/sign-up');
  };

  const homeHandler = () => {
    navigate('/');
  };

  const toggleUserPanel = () => {
    if (showPanel === 'd-none') {
      updatePanel('d-block');
    } else {
      updatePanel('d-none');
    }
  };

  if (userState.loading === true) {
    return (
      <div className="d-flex flex-column justify-center align-center w-100 h-100">
        <h1 className="session-title">Loading...</h1>
      </div>
    );
  }
  if (userState.user.logged_in === true && userState.loading === false) {
    return (
      <div className="trips-user-container d-flex" data-testid="app">
        <UserPanel display={showPanel} />
        <div>
          <div className="nav-bar d-flex space-between">
            <FontAwesomeIcon icon={faBars} onClick={toggleUserPanel} />
            <h1>Easy Trips</h1>
            <FontAwesomeIcon icon={faHome} onClick={homeHandler} />
          </div>
        </div>
        <Outlet />
      </div>
    );
  }
  return (
    <NoSession loginHandler={loginHandler} signUpHandler={signUpHandler} />
  );
}

export default App;
