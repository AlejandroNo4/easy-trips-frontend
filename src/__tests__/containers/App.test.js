import React from 'react';
import { Provider } from 'react-redux';
import {
  render,
  screen,
  cleanup,
  fireEvent,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import axiosMock from 'axios';
import { BrowserRouter } from 'react-router-dom';
import store from '../../store';
import App from '../../containers/App';

afterEach(cleanup);

const renderWithRedux = (component) => ({
  ...render(
    <Provider store={store}>
      <BrowserRouter>{component}</BrowserRouter>
    </Provider>,
  ),
});

describe('App Component', () => {
  beforeEach(() => {
    axiosMock.get.mockResolvedValueOnce({
      data: {
        favorites: [],
        logged_in: true,
        username: 'Test Expected Name',
        admin: true,
        email: 'test@mail.com',
        user_thumnail: [''],
        id: 1,
      },
    });
    renderWithRedux(<App />);
  });

  it('renders with Redux', async () => {
    const appComponent = await screen.findByTestId('app');
    expect(appComponent).toBeVisible();
  });

  it('matches the snapshot', async () => {
    const appComponent = await screen.findByTestId('app');
    expect(appComponent).toMatchSnapshot();
  });

  it('has the expected username', async () => {
    const appComponent = await screen.findByTestId('username');
    expect(appComponent).toHaveTextContent('Test Expected Name');
    expect(appComponent).not.toHaveTextContent('');
  });

  it('has the expected email', async () => {
    const appComponent = await screen.findByTestId('user-email');
    expect(appComponent).toHaveTextContent('test@mail.com');
    expect(appComponent).not.toHaveTextContent('');
  });

  it('has the expected thumnail when there is no url on the fetched data', async () => {
    const appComponent = await screen.findByTestId('user-thumnail');
    const noImg = 'https://cdn.pixabay.com/photo/2015/10/05/22/37/blank-profile-picture-973460_960_720.png';
    expect(appComponent).toHaveProperty('src', noImg);
  });

  it('Renders email and tranfers the value while typing', async () => {
    const button = await screen.findByTestId('fav-btn');
    expect(button).toBeVisible();
    fireEvent.click(button);
    expect(axiosMock.get).toHaveBeenCalledTimes(1);
    expect(axiosMock.get).not.toHaveBeenCalledTimes(0);
  });

  it('Renders email and tranfers the value while typing', async () => {
    const button = await screen.findByTestId('logout');
    expect(button).toBeVisible();
    fireEvent.click(button);
    axiosMock.delete.mockResolvedValueOnce({ data: { message: 'logged out' } });
    expect(axiosMock.delete).toHaveBeenCalledTimes(1);
    expect(axiosMock.delete).not.toHaveBeenCalledTimes(0);
  });
});
