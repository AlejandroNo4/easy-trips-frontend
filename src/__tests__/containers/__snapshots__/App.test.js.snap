// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component matches the snapshot 1`] = `
<div
  class="trips-user-container d-flex"
  data-testid="app"
>
  <div
    class="user-panel d-none"
  >
    <div
      class="user-info"
    >
      <img
        alt="Test Expected Name"
        class="thumnail bm-10"
        data-testid="user-thumnail"
        src="https://cdn.pixabay.com/photo/2015/10/05/22/37/blank-profile-picture-973460_960_720.png"
      />
      <div>
        <p
          class="bm-10 bold-text"
          data-testid="username"
        >
          Test Expected Name
        </p>
        <p
          class="bm-10 light-text"
          data-testid="user-email"
        >
          test@mail.com
        </p>
      </div>
    </div>
    <div>
      <div>
        <button
          class="btn-user-panel"
          data-testid="fav-btn"
          type="button"
        >
          Show favorites
        </button>
        <button
          class="btn-user-panel"
          type="button"
        >
          Update Account
        </button>
        <button
          class="btn-user-panel"
          type="button"
        >
          Create New Trip
        </button>
      </div>
      <div
        class="user-bottom-section"
      >
        <div
          class="user-line"
        />
        <button
          class="btn-user-panel"
          data-testid="logout"
          type="button"
        >
          Logout
        </button>
      </div>
    </div>
  </div>
  <div>
    <div
      class="nav-bar d-flex space-between"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-bars fa-w-14 "
        data-icon="bars"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
          fill="currentColor"
        />
      </svg>
      <h1>
        Easy Trips
      </h1>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-home fa-w-18 "
        data-icon="home"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 576 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M280.37 148.26L96 300.11V464a16 16 0 0 0 16 16l112.06-.29a16 16 0 0 0 15.92-16V368a16 16 0 0 1 16-16h64a16 16 0 0 1 16 16v95.64a16 16 0 0 0 16 16.05L464 480a16 16 0 0 0 16-16V300L295.67 148.26a12.19 12.19 0 0 0-15.3 0zM571.6 251.47L488 182.56V44.05a12 12 0 0 0-12-12h-56a12 12 0 0 0-12 12v72.61L318.47 43a48 48 0 0 0-61 0L4.34 251.47a12 12 0 0 0-1.6 16.9l25.5 31A12 12 0 0 0 45.15 301l235.22-193.74a12.19 12.19 0 0 1 15.3 0L530.9 301a12 12 0 0 0 16.9-1.6l25.5-31a12 12 0 0 0-1.7-16.93z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</div>
`;
